samples/batch-process-document.v1beta3.js

```diff
@@ -0,0 +1,156 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const uuid = require('uuid');
+
+async function main(
+  projectId = 'YOUR_PROJECT_ID',
+  location = 'YOUR_PROJECT_LOCATION',
+  processorId = 'YOUR_PROCESSOR_ID', // Create this in the Cloud Console
+  gcsInputUri = 'gs://cloud-samples-data/documentai/invoice.pdf',
+  gcsOutputUri = 'output-bucket',
+  gcsOutputUriPrefix = uuid.v4()
+) {
+  // [START documentai_batch_process_document]
+  /**
+   * TODO(developer): Uncomment these variables before running the sample.
+   */
+  // const projectId = 'YOUR_PROJECT_ID';
+  // const location = 'YOUR_PROJECT_LOCATION'; // Format is 'us' or 'eu'
+  // const processorId = 'YOUR_PROCESSOR_ID';
+  // const gcsInputUri = 'YOUR_SOURCE_PDF';
+  // const gcsOutputUri = 'YOUR_STORAGE_BUCKET';
+  // const gcsOutputUriPrefix = 'YOUR_STORAGE_PREFIX';
+
+  // Imports the Google Cloud client library
+  const {
+    DocumentProcessorServiceClient,
+  } = require('@google-cloud/documentai').v1beta3;
+  const {Storage} = require('@google-cloud/storage');
+
+  const clientOptions = {
+    apiEndpoint: 'us-documentai.googleapis.com',
+  };
+
+  const client = new DocumentProcessorServiceClient(clientOptions);
+  const storage = new Storage();
+
+  async function batchProcessDocument() {
+    const name = `projects/${projectId}/locations/${location}/processors/${processorId}`;
+
+    const inputConfig = {
+      gcsSource: gcsInputUri,
+      mimeType: 'application/pdf',
+    };
+    const inputConfigs = [inputConfig];
+
+    // Configure the batch process request.
+    const request = {
+      name,
+      inputConfigs,
+      outputConfig: {
+        gcsDestination: `${gcsOutputUri}/${gcsOutputUriPrefix}/`,
+      },
+    };
+
+    // Batch process document using a long-running operation.
+    // You can wait for now, or get results later.
+    // Note: first request to the service takes longer than subsequent
+    // requests.
+    try {
+      const [operation] = await client.batchProcessDocuments(request);
+
+      // Wait for operation to complete.
+      await operation.promise();
+    } catch (ex) {
+      // CURRENTLY the LRO returned from service has 'done = true' but
+      // does not provide a result. This raises an error from the google-gax
+      // library, despite the LRO actually succeeeding.
+      const lroError =
+        'Long running operation has finished but there was no result';
+
+      if (ex.message.indexOf(lroError) === -1) {
+        throw ex;
+      }
+    }
+
+    console.log('Document processing complete.');
+
+    // Query Storage bucket for the results file(s).
+    const query = {
+      prefix: gcsOutputUriPrefix,
+    };
+
+    console.log('Fetching results ...');
+
+    // List all of the files in the Storage bucket
+    const [files] = await storage.bucket(gcsOutputUri).getFiles(query);
+
+    files.forEach(async (fileInfo, index) => {
```

Comment:
So this is a bit of a confusing footgun.  As you've written this now, you are going to iterate over each file returned in the list, and create an async function.  If there are 10000 files, you are going to create 10000 concurrent requests, which will drown the node process and available handles in the OS.  I would suggest one of two routes:

## Do things serially
If you're ok with fetching each file serially, it's the most readable way to iterate over these, and ensure we don't crash:
```js
for (let i=0; i<files.length; i++) {
  const fileInfo = files[i];
  //...
}
```

## Fetch in parallel, but use a queue
This is a little more complicated.  If you really want to keep things parallel, you can drastically reduce the effect of a n unknown result set by using an async queue.  We tend to use [p-queue](https://www.npmjs.com/package/p-queue) for these purposes.  It looks something like this:
```js
const {default: Q} = require('p-queue');
 
const q = new Q({concurrency: 15});
await q.addAll(files.map(async fileInfo => { 
 // do stuff
});

// my syntax could be off here, going from memory
```

Reply:
I'd love to learn how to use that p-queue library. I'll give that a try!

Reply:
Done! PTAL at your earliest convenience.